pilot/pkg/model/endpointshards.go

```diff
@@ -92,6 +92,24 @@ func (es *EndpointShards) Keys() []ShardKey {
 	return keys
 }
 
+// CopyEndpoints takes a snapshot of all endpoints. As input, it takes a map of port name to number, to allow it to group
+// the results by service port number. This is a bit weird, but lets us efficiently construct the format the caller needs.
+func (es *EndpointShards) CopyEndpoints(portMap map[string]int) map[int][]*IstioEndpoint {
```

Comment:
Why do we use port name here, i am asking because it maybe empty

Reply:
if it's empty it's still valid though. you can only have 1 empty: 

> The name of this port within the service. This must be a DNS_LABEL. All ports within a ServiceSpec must have unique names. This maps to the 'Name' field in EndpointPort objects. Optional if only one ServicePort is defined on this service.

we have the same check in existing code

Reply:
port name is the only way to match an IstioEndpoint to a service you can see the same. in eds code